java/java.lsp.server/vscode/src/jdk/validation/project.ts

```diff
@@ -0,0 +1,171 @@
+}
+
+export async function getProjectFrom(projectUri: vscode.Uri): Promise<Project | undefined> {
+    const projectInfos: any[] = await vscode.commands.executeCommand(GET_PROJECT_INFO, projectUri.toString(), { projectStructure: true });
+    if (projectInfos?.length && projectInfos[0]) {
+        const projectDirectory = projectInfos[0].projectDirectory.toString();
+        const buildSystem: BuildSystemType = resolveBuildSystemType(projectUri, projectInfos[0].projectType);
+
+        switch (buildSystem) {
+            case BuildSystemType.MAVEN:
+                const mavenSubprojects: Project[] = projectInfos[0].subprojects
+                    .map((subproject: string) => new MavenProject(subproject, []))
+                return new MavenProject(projectDirectory, mavenSubprojects);
+            case BuildSystemType.GRADLE:
+                const gradleSubprojects: Project[] = projectInfos[0].subprojects
+                    .map((subproject: string) => new GradleProject(subproject, []))
+                return new GradleProject(projectDirectory, gradleSubprojects);
+            default:
+                break;
+        }
+    }
+    return Promise.resolve(undefined);
+}
+
+function resolveBuildSystemType(uri: vscode.Uri, projectType?: string): BuildSystemType {
+    if (projectType?.includes('gradle')) {
+        return BuildSystemType.GRADLE;
+    }
+    if (projectType?.includes('maven')) {
+        return BuildSystemType.MAVEN;
+    }
+    if (fs.existsSync(path.join(uri.fsPath, 'build.gradle'))) {
+        return BuildSystemType.GRADLE;
+    }
+    if (fs.existsSync(path.join(uri.fsPath, 'pom.xml'))) {
+        return BuildSystemType.MAVEN;
+    }
+    return BuildSystemType.UNKNOWN;
+}
+
+export abstract class Project {
+
+    readonly directory: string;
+    readonly subprojects: Project[];
+
+    constructor(directory: string, subprojects: any[]) {
+        this.directory = directory.replace('file:/', '');
+        this.subprojects = subprojects;
+    }
+
+    // Whether the project contains subprojects
+    containsSubprojects(): boolean {
+        return this.subprojects.length > 0;
+    }
+
+    async getJavaVersion(): Promise<number | undefined> {
+        if (!this.containsSubprojects()) {
+            return this.extractJavaVersion();
+        }
+
+        let maxJavaVersion: number | undefined;
+
+        for (const subproject of this.subprojects) {
+            const subInfos: any[] = await vscode.commands.executeCommand(GET_PROJECT_INFO, `file:/${subproject.directory}`);
+            if (subInfos?.length && subInfos[0]) {
+                const javaVersion = subproject.extractJavaVersion();
+
+                if (!maxJavaVersion || (javaVersion && javaVersion > maxJavaVersion)) {
+                    maxJavaVersion = javaVersion;
+                }
+            }
+        }
+        return maxJavaVersion;
+    }
+
+    // Extracts project java version
+    protected abstract extractJavaVersion(): number | undefined
```

Comment:
Please file an issue against Netbeans project, to extend `GET_PROJECT_INFO` (or provide an additional command) with java settings -- the source/target/release info is already internally available from maven/gradle parsing itself; just not exported to LS protocol.

Note the issue # here and note this code location in that issue as well, so it is eventually updated when the info becomes available.